File: Gatito Enojado - Python + Pygame/src/player.py
"""
Módulo del Jugador
Gestiona el movimiento del personaje, animaciones por fotogramas, colisiones y estado de vida/invulnerabilidad.
"""

import os
import sys

import pygame

from src.constants import (
    MAP_OFFSET_X,
    MAP_OFFSET_Y,
    PLAYER_LIVES,
    PLAYER_SPEED,
    SCREEN_HEIGHT,
    SCREEN_WIDTH,
)


def resource_path(relative_path):
    """Obtiene la ruta absoluta segura para las animaciones del jugador."""
    try:
        base_path = sys._MEIPASS
    except Exception:  # noqa: BLE001
        base_path = os.path.abspath(".")
    return os.path.join(base_path, relative_path)

class Player(pygame.sprite.Sprite):
    def __init__(self, x, y):
        super().__init__()
        self.rect = pygame.Rect(x, y, 28, 28)
        # Carga de frames para la animación de caminata
        self.base_images = [
            pygame.image.load(resource_path("assets/player_0.png")).convert_alpha(),
            pygame.image.load(resource_path("assets/player_1.png")).convert_alpha(),
            pygame.image.load(resource_path("assets/player_2.png")).convert_alpha()
        ]
        self.image_index = 0
        self.animation_timer = 0
        self.facing_left = False
        self.poops = []
        self.is_alive = True
        self.speed = PLAYER_SPEED
        self.lives = PLAYER_LIVES
        self.invulnerable_until = 0

    def update(self, keys, walls, destructibles, poops):
        """Actualiza la posición del jugador en base a las teclas presionadas y maneja colisiones eje por eje."""
        if not self.is_alive: 
            return
            
        dx, dy = 0, 0
        moving = False
        
        # Lectura de controles direccionales
        if keys[pygame.K_LEFT]:
            dx = -self.speed 
            moving = True 
            self.facing_left = True
        elif keys[pygame.K_RIGHT]:
            dx = self.speed
            moving = True 
            self.facing_left = False
            
        if keys[pygame.K_UP]:
            dy = -self.speed 
            moving = True
        elif keys[pygame.K_DOWN]:
            dy = self.speed 
            moving = True
            
        # Control de ciclo de animación al caminar
        if moving:
            self.animation_timer += 1
            if self.animation_timer >= 10:
                self.animation_timer = 0 
                self.image_index = (self.image_index + 1) % 3
        else:
            self.image_index = 0 
            self.animation_timer = 0

        # Filtro de obstáculos sólidos (las trampas recién colocadas son traspasables brevemente)
        current_time = pygame.time.get_ticks()
        solid_poops = [p.rect for p in poops if (current_time - p.spawn_time) > 200]
        obstacles = walls + destructibles + solid_poops
        
        # Movimiento horizontal y resolución de colisiones en eje X
        self.rect.x += dx
        for obs in obstacles:
            if self.rect.colliderect(obs):
                if dx > 0: 
                    self.rect.right = obs.left
                if dx < 0: 
                    self.rect.left = obs.right
                    
        # Movimiento vertical y resolución de colisiones en eje Y
        self.rect.y += dy
        for obs in obstacles:
            if self.rect.colliderect(obs):
                if dy > 0: 
                    self.rect.bottom = obs.top
                if dy < 0: 
                    self.rect.top = obs.bottom
        
        # Restricción de movimiento dentro de los límites del mapa
        area = pygame.Rect(MAP_OFFSET_X, MAP_OFFSET_Y, SCREEN_WIDTH - 2 * MAP_OFFSET_X, SCREEN_HEIGHT - 2 * MAP_OFFSET_Y)
        self.rect.clamp_ip(area)

    def die(self):
        """Resta una vida al jugador aplicando un periodo de invulnerabilidad temporal."""
        current_time = pygame.time.get_ticks()
        if current_time > self.invulnerable_until:
            self.lives -= 1 
            self.invulnerable_until = current_time + 1000
            if self.lives <= 0: 
                self.is_alive = False
            return True
        return False

    def draw(self, screen):
        """Dibuja al jugador aplicando efecto de parpadeo si se encuentra en estado de invulnerabilidad."""
        if self.is_alive:
            if pygame.time.get_ticks() % 200 < 100 and pygame.time.get_ticks() < self.invulnerable_until: 
                return
            img = self.base_images[self.image_index]
            display_image = pygame.transform.flip(img, self.facing_left, False)
            screen.blit(display_image, self.rect)

File: js/player.js
import { MAP_OFFSET_X, MAP_OFFSET_Y, SCREEN_WIDTH, SCREEN_HEIGHT, PLAYER_SPEED } from './constants.js';

export class Player {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = 28;
        this.height = 28;
        this.speed = PLAYER_SPEED;
        this.lives = 3;
        this.isAlive = true;
        this.facingLeft = false;
        
        this.images = [0, 1, 2].map(i => {
            const img = new Image();
            img.src = `assets/player_${i}.png`;
            return img;
        });
        this.imageIndex = 0;
        this.animationTimer = 0;
        this.invulnerableUntil = 0;
    }

    update(keys, obstacles) {
        if (!this.isAlive) return;

        let dx = 0;
        let dy = 0;
        let moving = false;

        if (keys['ArrowLeft']) { dx = -this.speed; moving = true; this.facingLeft = true; }
        else if (keys['ArrowRight']) { dx = this.speed; moving = true; this.facingLeft = false; }

        if (keys['ArrowUp']) { dy = -this.speed; moving = true; }
        else if (keys['ArrowDown']) { dy = this.speed; moving = true; }

        if (moving) {
            this.animationTimer++;
            if (this.animationTimer >= 10) {
                this.animationTimer = 0;
                this.imageIndex = (this.imageIndex + 1) % 3;
            }
        } else {
            this.imageIndex = 0;
            this.animationTimer = 0;
        }

        // Movimiento en X y colisiones
        this.x += dx;
        let playerRect = this.getRect();
        for (let obs of obstacles) {
            if (this.checkCollision(playerRect, obs)) {
                if (dx > 0) this.x = obs.x - this.width;
                if (dx < 0) this.x = obs.x + obs.width;
                playerRect = this.getRect();
            }
        }

        // Movimiento en Y y colisiones
        this.y += dy;
        playerRect = this.getRect();
        for (let obs of obstacles) {
            if (this.checkCollision(playerRect, obs)) {
                if (dy > 0) this.y = obs.y - this.height;
                if (dy < 0) this.y = obs.y + obs.height;
                playerRect = this.getRect();
            }
        }

        const minX = MAP_OFFSET_X;
        const minY = MAP_OFFSET_Y;
        const maxX = MAP_OFFSET_X + 25 * 32 - this.width;
        const maxY = MAP_OFFSET_Y + 18 * 32 - this.height;

        if (this.x < minX) this.x = minX;
        if (this.x > maxX) this.x = maxX;
        if (this.y < minY) this.y = minY;
        if (this.y > maxY) this.y = maxY;
    }

    getRect() {
        return { x: this.x, y: this.y, width: this.width, height: this.height };
    }

    checkCollision(rect1, rect2) {
        return rect1.x < rect2.x + rect2.width &&
               rect1.x + rect1.width > rect2.x &&
               rect1.y < rect2.y + rect2.height &&
               rect1.y + rect1.height > rect2.y;
    }

    die(currentTime) {
        if (currentTime > this.invulnerableUntil) {
            this.lives--;
            this.invulnerableUntil = currentTime + 1000;
            if (this.lives <= 0) this.isAlive = false;
            return true;
        }
        return false;
    }

    draw(ctx, currentTime) {
        if (!this.isAlive) return;
        if (currentTime < this.invulnerableUntil && Math.floor(currentTime / 100) % 2 === 0) {
            return; 
        }

        ctx.save();
        if (this.facingLeft) {
            ctx.scale(-1, 1);
            ctx.drawImage(this.images[this.imageIndex], -this.x - this.width, this.y, this.width, this.height);
        } else {
            ctx.drawImage(this.images[this.imageIndex], this.x, this.y, this.width, this.height);
        }
        ctx.restore();
    }
}